### ISSUE TITLE
Better error reporting of matrix conformability

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#22]
Running a model ("m1") with an error in the algebra (attached as zipped m1.Rda)) model,  get the following error:

    Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings,  : 
      Non-conformable matrices [(1 x 2) and (1 x 2)] in Matrix Multiply.

Be great to know which matrices... in which models...

[m1.rda.zip](https://github.com/OpenMx/OpenMx/files/1246475/m1.rda.zip)

**Comment by mcneale** [comment]
I whole-heartedly agree!  Sometimes it isn’t easy to do this because the matrices in question don’t exist as mxMatrices, but are the results of prior algebras.  However, it seems that our error message is dumb in both contexts, and should be fixed wherever possible.  If we were really smart, we’d say what algebra produced the intermediate results causing the problem, but that’s a bigger project.

> xalg <- mxAlgebra(xx %*% xx, name="xxxx")
> xx <- mxMatrix("Full", 2, 1, name="xx")
> xmod <- mxModel("Test algebra", xx, xalg)
> mxRun(xmod)
Running Test algebra with 0 parameters
Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings,  :
  Non-conformable matrices [(2 x 1) and (2 x 1)] in Matrix Multiply.

> mxVersion()
OpenMx version: 2.7.16 [GIT v2.7.16]
R version: R version 3.4.0 (2017-04-21)
Platform: x86_64-apple-darwin15.6.0
MacOS: 10.12.6
Default optimiser: NPSOL
NPSOL-enabled?: Yes
OpenMP-enabled?: Yes

Cheers
Mike

Michael C. Neale, Ph.D.
Professor, Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Shipping Address: 800 East Leigh St. Suite 1-156, Richmond VA 23219-1534
Mailing Address : Box 980126 Richmond VA 23298-0126
Telephone: (804) 828-3369  michael.neale@vcuhealth.org<mailto:michael.neale@vcuhealth.org>

On Aug 23, 2017, at 1:57 PM, Tim Bates <notifications@github.com<mailto:notifications@github.com>> wrote:


Running a model ("m1") with an error in the algebra (attached as zipped m1.Rda)) model, get the following error:

Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings,  :
  Non-conformable matrices [(1 x 2) and (1 x 2)] in Matrix Multiply.


Be great to know which matrices... in which models...

m1.rda.zip<https://github.com/OpenMx/OpenMx/files/1246475/m1.rda.zip>

**Comment by mcneale** [comment]
That’s great!  I note that it works for addition very nicely as well, even when one or more arguments is the result of an expression.

However, is there a mis-ordering of the subsequent argument descriptors?  Test algebra.xx is not the one with the transpose, it is simply matrix(c(    # 1x2   0,   0), byrow=TRUE, nrow=1, ncol=2), and ‘? arg 1’ is the one that has been transposed.



> xx <- mxMatrix("Full", 2, 1)
> xalg <- mxAlgebra(xx + t(xx), name="xxxx")
> xmod <- mxModel("Test algebra", xx, xalg)
> mxRun(xmod)
Running Test algebra with 0 parameters
Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings,  :
  Element-wise '+' not conformable: 'Test algebra.xx' is 2x1 and '? arg 1' is 1x2
Test algebra.xx = t( matrix(c(    # 2x1
   0,   0), byrow=TRUE, nrow=1, ncol=2))
? arg 1 =  matrix(c(    # 1x2
   0,   0), byrow=TRUE, nrow=1, ncol=2)


The only other comment I would make is that it would, I think, be better still if the lines were not broken in the middle of the algebra:

Test algebra.xx = t( matrix(c(    # 2x1   0,   0), byrow=TRUE, nrow=1, ncol=2))
? arg 1 =  matrix(c(    # 1x2   0,   0), byrow=TRUE, nrow=1, ncol=2)
would be easier to read except with very long expressions.


Cheers
Mike

Michael C. Neale, Ph.D.
Professor, Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Shipping Address: 800 East Leigh St. Suite 1-156, Richmond VA 23219-1534
Mailing Address : Box 980126 Richmond VA 23298-0126
Telephone: (804) 828-3369  michael.neale@vcuhealth.org<mailto:michael.neale@vcuhealth.org>

On Aug 24, 2017, at 10:19 AM, Joshua Pritikin <notifications@github.com<mailto:notifications@github.com>> wrote:


In commit 1dded05<https://github.com/OpenMx/OpenMx/commit/1dded05396852c5f7750e9e0065ceb64be1b741a>, I improved it to,

Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings, : Non-conformable matrices 'Test algebra.xx' 2x1 and 'Test algebra.xx' 2x1 in matrix multiply

This change only affects matrix multiplication. Error messages from other binary ops are not affected.

**Comment by RMKirkpatrick** [comment]
I guess it'd be nice to know which matrices are at fault in the case of non-conformability for other operators--addition & subtraction, elementwise multiplication & division, elementwise powering, etc.

> The only other comment I would make is that it would, I think, be better still if the lines were not broken in the middle of the algebra:

I disagree.  It only looks odd because the matrix in question is really small.  This is how we print matrices from the backend, and in general, it's a sane way to do so, and I'm speaking from ample experience with reading verbose backend printing.

**Comment by RMKirkpatrick** [comment]
P.S. See mxPrintMat(), defined starting line 509 of src/omxMatrix.h .

**Comment by mcneale** [comment]
No, I mean what is attached to ? arg 1 is not the algebra that generated it, which it should be, right?

Instead, Test algebra.xx which should just be a matrix comes out as a formula with t()
Test algebra.xx = t( matrix(c(    # 2x1
   0,   0), byrow=TRUE, nrow=1, ncol=2))

and ? arg 1 is printed as if it was just a vanilla matrix

? arg 1 =  matrix(c(    # 1x2
   0,   0), byrow=TRUE, nrow=1, ncol=2)

whereas, surely, it should be the formula?

> xx <- mxMatrix("Full", 2, 1)
> xalg <- mxAlgebra(xx + t(xx), name="xxxx")
> xmod <- mxModel("Test algebra", xx, xalg)
> mxRun(xmod)
Running Test algebra with 0 parameters
Error in runHelper(model, frontendStart, intervals, silent, suppressWarnings,  :
  Element-wise '+' not conformable: 'Test algebra.xx' is 2x1 and '? arg 1' is 1x2
Test algebra.xx = t( matrix(c(    # 2x1
   0,   0), byrow=TRUE, nrow=1, ncol=2))
? arg 1 =  matrix(c(    # 1x2
   0,   0), byrow=TRUE, nrow=1, ncol=2)


Cheers
Mike

Michael C. Neale, Ph.D.
Professor, Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Shipping Address: 800 East Leigh St. Suite 1-156, Richmond VA 23219-1534
Mailing Address : Box 980126 Richmond VA 23298-0126
Telephone: (804) 828-3369  michael.neale@vcuhealth.org<mailto:michael.neale@vcuhealth.org>

On Aug 24, 2017, at 11:00 AM, Joshua Pritikin <notifications@github.com<mailto:notifications@github.com>> wrote:


The order looks fine. ‘? arg 1’ is the name of the transpose result implicitly created matrix.

I guess it'd be nice to know which matrices are at fault in the case of non-conformability for other operators--addition & subtraction, elementwise multiplication & division, elementwise powering, etc.

I think we have similar messages for all binary ops.

**Comment by jpritikin** [comment]
I suppose we could stringify the abstract syntax tree to some depth limit to provide more context, but not to arbitrary depth. For example, suppose I write `xx * t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(t(xx)))))))))))))` or some bulkier expression. The error message could get too long.

**Comment by jpritikin** [comment]
See omxMatrix.h line 462
